### PR TITLE
CKAN / DataPackages / JSON Table Schema integration spec

### PR DESCRIPTION
**Pull request by amercader**
Rendered doc:

https://github.com/ckan/ideas-and-roadmap/blob/datapackages-spec/specs/datapackages/README.md

### CODE REVIEW

File: specs/datapackages/README.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+1. Allow importing Data Packages into CKAN, creating the necessary datasets and resources, both via the UI and the API
+
+2. Allow exporting CKAN datasets as Data Packages, regardless of whether they were imported from Data Packages in a first instance or not.
+
+3. Support schema definition for CKAN resources, allowing publishers to describe the contents of published files
+
+
+For achieving this it seems sensible to reuse as much stuff as we can from ckanext-datapackager, extending it if necessary
+
+The first two points could be implemented either in core or separate extension, but I'd like to propose that the third regarding resource schemas is implemented in core (in a generic form). It seems central enough to all future work around data import, cleaning, etc (eg closer integration with DataStore) to justify it.
```
Comment (rossjones) [inline specs/datapackages/README.md]: s/The first two points could be implemented either in core or separate extension/The first two points should be implemented as a core extension/

File: specs/datapackages/README.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
 characters. It will function as a unique identifier and therefore SHOULD be unique in relation to any registry in which this package will be deposited (and preferably globally unique).
+
+We can not assume that these identifiers will be unique and that they will not clash with existing CKAN datasets. We can append stuff to the name but then we need to consider what happens when we reupload the same Data Package (ie to update the dataset).
+
+**Q:** Do we upload the files described on the Data Package to the CKAN server?
+I would say not by default. As long as the files are publicly accessible the DataPusher can get them into the DataStore.
+
+**Q:** Do we keep a copy of the original `datapackage.json` ?
+I think yes. Trying to store all fields as extras and replicate them back when generating the datapackage back would be too much hassle. Also we would need to keep adding fields as the spec changes and handle arbitrary fields added by publishers. An easier approach might be outputting back what was uploaded, but with the fields that are actually mapped to CKAN (title, description, keywords etc) updated, as these might have been modified in CKAN.
+
```
Comment (rossjones) [inline specs/datapackages/README.md]: Is there an ETA on when the spec will be 1.0, and no longer a draft? Having the spec shifting under the feet of the developer seems like it will be problematic, and force things like storing an updated datapackage.json unnecessarily?

File: specs/datapackages/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+Accessing `/dataset/{id}/datapackage.json` will output the generated Data Package for this dataset. Similarly as above, we'll have a `datapackage_export` action (with the same auth as `package_show`) and a controller to handle requests.
+
+### Potential issues
+
+See point above regarding outputting back custom fields.
+
+I think that the JSON files could be generated on the fly, but if performance is an issue the Data Package JSON blobs for each dataset could be generated before hand (using `after_create` and `after_update`) and stored in the DB or even as a field in the Solr index.
+
+**Q:** So far we've only considered giving back the `datapackage.json` file, but would it make sense to have an extra endpoint that returns a zip of the `datapackage.json` plus the data files?
```
Comment (rossjones) [inline specs/datapackages/README.md]: A lot of the zipping code is available at https://github.com/datagovuk/ckanext-packagezip

File: specs/datapackages/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+* `sources`
+* `license`
+
+This is not as problematic as dataset fields, as we can add arbitrary top level fields to resources.
+
+## Importing Data Packages
+
+### Implementation
+
+At a higher level this seems quite straight-forward: a  `datapackage_import` or similar action which inherits the auth from `package_create` and `package_update` (eg you need to belong to an org by default etc). The action will validate the JSON object, transform it to a CKAN dict and create or update the relevant package. A controller will handle requests from the UI (at `/datapackage/import`).
```
Comment (davidread) [inline specs/datapackages/README.md]: We should avoid one-off imports of metadata, because unless you have ongoing updates then you now have two diverging copies of the metadata. Rather than a 'datapackage_import' action, at the very least we should have datapackage_create and datapackage_update. But more-over I'd prefer data packages to instead be harvested, so that you keep the CKAN copy in sync with the source of the metadata. Datasets that don't have updates are dead, and are the minority, so let's focus on use cases for datasets which do have updates.

File: specs/datapackages/README.md (hunk 2 above)
Comment (pwalsh) [inline specs/datapackages/README.md]: Hi @rossjones 

where does it say the spec is a draft?

* http://dataprotocols.org/data-packages/index.html
* http://dataprotocols.org/tabular-data-package/index.html

The Base Data Package spec, and Tabular Data Package, are each currently beta of v1. I don't see any great obstacle to removing the beta label very soon, if that is the issue.

File: specs/datapackages/README.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+Identifiers: Data Packages are solely identified by the `name` field, which according to the Data Package spec:
+
+> [is a] short url-usable (and preferably human-readable) name of the package. This MUST be lower-case and contain only alphanumeric characters along with 
., 
_ or 
-
 characters. It will function as a unique identifier and therefore SHOULD be unique in relation to any registry in which this package will be deposited (and preferably globally unique).
+
+We can not assume that these identifiers will be unique and that they will not clash with existing CKAN datasets. We can append stuff to the name but then we need to consider what happens when we reupload the same Data Package (ie to update the dataset).
```
Comment (davidread) [inline specs/datapackages/README.md]: We could assign a data package a guid (used for matching when you reimport it) which contains its name and author. So the author becomes its namespace. And then in ckan we just munge the name to ensure its unique within CKAN.

File: specs/datapackages/README.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+**Q:** Do we upload the files described on the Data Package to the CKAN server?
+I would say not by default. As long as the files are publicly accessible the DataPusher can get them into the DataStore.
+
+**Q:** Do we keep a copy of the original `datapackage.json` ?
+I think yes. Trying to store all fields as extras and replicate them back when generating the datapackage back would be too much hassle. Also we would need to keep adding fields as the spec changes and handle arbitrary fields added by publishers. An easier approach might be outputting back what was uploaded, but with the fields that are actually mapped to CKAN (title, description, keywords etc) updated, as these might have been modified in CKAN.
```
Comment (davidread) [inline specs/datapackages/README.md]: Agree, just like we already do with the harvester already.

File: specs/datapackages/README.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+## Resource schemas
+
+
+### Implementation
+
+Add new core fields to the resource model:
+    * `schema`: a JSON object describing the structure of the resource file, or a URL linking to the schema.
+    * `schemaType`: a string or URI that specifies what is the specification for the schema. 
+    * `schemaVersion`: the version of the spec being used
```
Comment (davidread) [inline specs/datapackages/README.md]: Yep, I'm keen on having all three of these and the latter two are optional.

I also agree on allowing schema to be either JSON or a URL - both have plenty of use cases in my mind.

File: specs/datapackages/README.md (hunk 2 above)
Comment (rossjones) [inline specs/datapackages/README.md]: In the language section..

```
    NOTE: This is a draft specification and still under development.
```

I think having a concrete version that won't change during development would be beneficial. Bug fixes can come in point releases I guess.

File: specs/datapackages/README.md (hunk 7 above)
Comment (davidread) [inline specs/datapackages/README.md]: The names are fine, although I'm not sure about the camelcase - maybe `schema_type` and `schema_version` to match things all the others in CKAN like `owner_org`?

File: specs/datapackages/README.md (hunk 5 above)
Comment (amercader) [inline specs/datapackages/README.md]: That sounds good, although perhaps I'd use the organization name if possible. I'd assume that if two users on the same org upload a `budget-2016` file they would be working on the same one (and if not they should use different names)

File: specs/datapackages/README.md (hunk 3 above)
Comment (amercader) [inline specs/datapackages/README.md]: Great, ckanext-datapackager also has some.

File: specs/datapackages/README.md (hunk 1 above)
Comment (amercader) [inline specs/datapackages/README.md]: Do you mean an extension shipped with CKAN core, like `datastore`, `recline_view` etc? If so I'm happy with it.

File: specs/datapackages/README.md (hunk 4 above)
Comment (amercader) [inline specs/datapackages/README.md]: I can see how keeping copies in sync might be an issue, and we should totally write a Data Packages harvester for those who want to use it, but having to set up harvesting on the CKAN side and configure a harvest source on the data publisher side, seems like a quite steep barrier of entry for publishing Data Packages.

The main use case I have in mind is a data hacker that has been working on their datasets and wants to publish them online to CKAN either:

* directly via an upload form
* via a command line tool like `dpm` or similar

Basically a "push" scenario rather than CKAN pulling. For handling updates either you reupload (and by having a guid we ensure we update the correct dataset) or configure your client app to push periodically or on demand (I'd imagine that they would support something like `dpm push`.

By `datapackage_create` and `datapackage_update` do you mean something like `create_dataset_from_datapackage` and `update_dataset_from_datapackage`? `datapackage_create` seems to imply that we are creating a Data Package, not a CKAN dataset.

File: specs/datapackages/README.md (hunk 7 above)
Comment (amercader) [inline specs/datapackages/README.md]: Yeah, absolutely case should be as you say

File: specs/datapackages/README.md (hunk 1 above)
Comment (rossjones) [inline specs/datapackages/README.md]: Exactly like those, yes.

**Comment by danmihaila** [comment]
I find also very useful to have a generated"datapackage.json" for each resource in case needed. It is not very clear where the information that could be generated on the fly (datapackage.json) will be stored.

File: specs/datapackages/README.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+Perhaps to make it more accessible we can link to this page from the dataset creation form (from the top bar or the left hand column).
+
+### Potential issues
+
+Identifiers: Data Packages are solely identified by the `name` field, which according to the Data Package spec:
+
+> [is a] short url-usable (and preferably human-readable) name of the package. This MUST be lower-case and contain only alphanumeric characters along with ., _ or - characters. It will function as a unique identifier and therefore SHOULD be unique in relation to any registry in which this package will be deposited (and preferably globally unique).
+
+We can not assume that these identifiers will be unique and that they will not clash with existing CKAN datasets. We can append stuff to the name but then we need to consider what happens when we reupload the same Data Package (ie to update the dataset).
```
Comment (rossjones) [inline specs/datapackages/README.md]: As we can't assume that the name is unique, is this a flaw with the specification that may need fixing?  Perhaps using the URL of the datapackage as the unique identifier might be better?

Comment (pwalsh) [inline specs/datapackages/README.md]: I think it needs addressing in the spec itself. Also see discussion in https://github.com/dataprotocols/dataprotocols/issues/220

File: specs/datapackages/README.md (hunk 2 above)
Comment (pwalsh) [inline specs/datapackages/README.md]: This is simple to resolve. We can move to a v1 stable very soon.

File: specs/datapackages/README.md (hunk 3 above)
Comment (pwalsh) [inline specs/datapackages/README.md]: Also note that we are looking to formalize zipping: https://github.com/dataprotocols/dataprotocols/pull/198

Comment (rossjones) [inline specs/datapackages/README.md]: :+1:

File: specs/datapackages/README.md (hunk 4 above)
Comment (davidread) [inline specs/datapackages/README.md]: I guess the principle here is that the metadata should be kept close to the data. So if a hacker or organization publishes both then you should push both or pull both. I can see there's also the case where a hacker creates metadata that points to an official source, so it's not possible to require they come via the same method. So you're write, for the case of data sources like hackers or organizations that choose to make the CKAN their primary place of distributing data, allowing upload of a data package makes some sense.

And yes I see from other bits of your doc that you intended `datapackage_import` to do both create and update, as appropriate, so that's fine.

Comment (vitorbaptista) [inline specs/datapackages/README.md]: @amercader and me talked about this today. We could see two options to update an existing dataset with a datapackage:

1. Have a unique ID in the datapackage.json that we can store in the dataset, so we keep the relationship between them;
2. Making the user explicitly ask to update a certain dataset.

The first solution might result in a better UX (depending on who you ask), but it complicates things. We would have to change the data package schema to include a globally-unique identifier, and maybe handle cases where they don't exist (i.e. on datapackages using an older version of the schema).

Using the second solution, we would always create a new dataset when importing a datapackage.json. If the user wants to update a current dataset, she would have to go to the dataset's edit page and click in an "import" link. We could even add a "Import to Dataset" field on the datapackage import page that defaults to "Create new dataset", but can be changed to point whatever dataset the user wants to import into.

@amercader and I preferred the second solution, as it's explicit, doesn't require changes to the datapackage schema, and is easier to implement.

What do you guys think?

Comment (vitorbaptista) [inline specs/datapackages/README.md]: Still about updates: what happens when you import a datapackage.json into an existing CKAN dataset? Does the data in the datapackage.json takes precedence over the data in CKAN? What if there are more resources in CKAN than in the datapackage.json, should they be deleted?

I feel that importing a datapackage.json into a CKAN dataset means:

1. Drop everything from the CKAN dataset, including the resources;
2. Import from the datapackage.json.

The resulting dataset created when importing into an existing dataset or importing into a new one should be the same. In other words, we're not merging the datapackage.json into a CKAN dataset, but overwritting it.

Thoughts?

Comment (davidread) [inline specs/datapackages/README.md]: Asking the user whether the import of data package should create a new dataset or replace a nominated one seems a bit clumsy. And is no good for harvesting, which I think is a pretty major use case.

I don't have much time for any metadata standard that doesn't allow identification of which record is which by having an unchanging unique identifier. I understand that it is a bit of a faff to pick a unique number for a record, but it's absolutely essential to keep track of these things when you have more than a handful.

Regarding whether to replace the whole record or not, I think that is a reasonable start. However in DGU there are some extra fields that we'd want to keep generally. e.g. the publisher may manually assign the dataset a theme. Or drop it into a group specific to the particular CKAN. Or QA may fill in a resource format that was not set. We have this issue with harvesting too. How about you keep dataset's groups, and allow in the CKAN admin to specify which extra fields are left alone too.

Comment (wardi) [inline specs/datapackages/README.md]: :+1: from me on a unique id in the datapackage format. It's just so much harder to work with data when you can't tell when you're updating or creating a record.

Comment (amercader) [inline specs/datapackages/README.md]: > Asking the user whether the import of data package should create a new dataset or replace a nominated one seems a bit clumsy. And is no good for harvesting, which I think is a pretty major use case.

Just to clarify option 2 in the first comment didn't imply actually asking the user, just having two separate places for creating (`/dataset/import/datapackage` or whatever) and updating (`/dataset/{id}/edit/datapackage`) datasets from datapackage. Much in the same way that you create ( `/dataset/new`) or update  (`/dataset/{id}/edit`) datasets.

But I agree that having a global id is the only sane way going forward to cover cases like API, harvesting etc. We should store the datapackage id as the `guid` extra like most of the harvesters do. And for datapackages that are generated from a CKAN datasets we assign the internal CKAN id as datapackage id.

This will require a change on the datapackage spec though, as discussed in dataprotocols/dataprotocols#220

**Comment by vitorbaptista** [comment]
I've started changing the old https://github.com/ckan/ckanext-datapackager extension to work on CKAN 2.4 and implement the ideas discussed here about importing/exporting datapackages into CKAN. I've created a milestone to track this on https://github.com/ckan/ckanext-datapackager/milestones/Importing%20and%20Exporting%20Data%20Packages%20on%20CKAN%202.4.

I'm documenting the issues I'm encountering as I work on it. For example, I've written about the mapping between CKAN fields to datapackage fields at https://github.com/ckan/ckanext-datapackager/issues/25#issuecomment-161761510. Another issue was how to deal with the extras fields (https://github.com/ckan/ckanext-datapackager/issues/27).

If you guys have some time, please take a look :+1:
